ssl/quic/quic_txp.c

```diff
@@ -0,0 +1,2027 @@
+        have_pkt_for_el[enc_level] = txp_el_pending(txp, enc_level, archetype);
+        if (have_pkt_for_el[enc_level])
+            ++num_el_in_dgram;
+    }
+
+    if (num_el_in_dgram == 0)
+        return 1;
+
+    /*
+     * Should not be needed, but a sanity check in case anyone else has been
+     * using the QTX.
+     */
+    ossl_qtx_finish_dgram(txp->args.qtx);
+
+    for (enc_level = QUIC_ENC_LEVEL_INITIAL;
+         enc_level < QUIC_ENC_LEVEL_NUM;
+         ++enc_level) {
+        if (!have_pkt_for_el[enc_level])
+            continue;
+
+        is_last_in_dgram = (pkts_done + 1 == num_el_in_dgram);
+        rc = txp_generate_for_el(txp, enc_level, archetype, is_last_in_dgram,
+                                 have_pkt_for_el[QUIC_ENC_LEVEL_INITIAL]);
+
+        if (rc != TXP_ERR_SUCCESS) {
+            /*
+             * If we already successfully did at least one, make sure we report
+             * this via the return code.
+             */
+            if (pkts_done > 0)
+                break;
+            else
+                return 0;
+        }
+
+        ++pkts_done;
+    }
+
+    ossl_qtx_finish_dgram(txp->args.qtx);
+    return 2;
+}
+
+struct archetype_data {
+    unsigned int allow_ack                  : 1;
+    unsigned int allow_ping                 : 1;
+    unsigned int allow_crypto               : 1;
+    unsigned int allow_handshake_done       : 1;
+    unsigned int allow_path_challenge       : 1;
+    unsigned int allow_path_response        : 1;
+    unsigned int allow_new_conn_id          : 1;
+    unsigned int allow_retire_conn_id       : 1;
+    unsigned int allow_stream_rel           : 1;
+    unsigned int allow_conn_fc              : 1;
+    unsigned int allow_conn_close           : 1;
+    unsigned int allow_cfq_other            : 1;
+    unsigned int allow_new_token            : 1;
+    unsigned int allow_force_ack_eliciting  : 1;
+};
+
+static int txp_get_archetype_data(uint32_t enc_level,
+                                  uint32_t archetype,
+                                  struct archetype_data *a)
+{
+    char is_0  = (enc_level == QUIC_ENC_LEVEL_0RTT);
+    char is_1  = (enc_level == QUIC_ENC_LEVEL_1RTT);
+    char is_01 = (is_0 || is_1);
+    char is_ih1
+        = (enc_level == QUIC_ENC_LEVEL_INITIAL
+           || enc_level == QUIC_ENC_LEVEL_HANDSHAKE
+           || is_1);
+
+    a->allow_ack                    = is_ih1;
+    a->allow_ping                   = 1;
+    a->allow_crypto                 = is_ih1;
+    a->allow_handshake_done         = is_1;
+    a->allow_path_challenge         = 0;
+    a->allow_path_response          = 0;
+    a->allow_new_conn_id            = is_01;
+    a->allow_retire_conn_id         = is_01;
+    a->allow_stream_rel             = is_01;
+    a->allow_conn_fc                = is_01;
+    a->allow_conn_close             = 1;
+    a->allow_cfq_other              = !is_0;
+    a->allow_new_token              = is_1;
+    a->allow_force_ack_eliciting    = !is_0;
```

Comment:
This all seems like fixed unchanging data for any given enc_level. Would a table lookup be better rather than calculating it every time?

Reply:
Maybe. I'm not totally convinced this is going to make a significant performance difference though (plus, recalculating data can be faster than table lookups if the latter would mean a cache miss), and it seems premature to optimise at this stage. I'd like to keep this as this is until we have requirements for packet generation locked down; some of this might need tweaking.